### ISSUE TITLE
GWT/html5 build - dangling entries in the joint list of a box2d world

### ISSUE THREAD

**Issue by Snack-a-Dog** [libgdx/libgdx#3661]
In the html5 build, the internal joints list of the World class contains dangling entries after a body with multiple attached joints was destroyed. The method ```World.getJointCount()``` returns a different value than the size of the list returned by ```World.getJoints()```. The list still contains joints that should have been destroyed together with the body that they were attached to.

Example code (please paste as an ApplicationListener into a libGDX project):
```java
public class PhysicsDebug extends ApplicationAdapter {
	
    private World world;
    
	@Override
	public void create () {
	    world = new World(new Vector2(0, 0),  true);
	    
	    Body bodyA = createBody();
	    Body bodyB = createBody();
	    createDistanceJointBetweenBodies(bodyA, bodyB);
	    createDistanceJointBetweenBodies(bodyA, bodyB);
	    
	    Gdx.app.setLogLevel(Application.LOG_DEBUG);
	    
	    Gdx.app.log("debug", "Created the bodies and the joints");
	    Gdx.app.log("debug", "world.getJointCount() -> " + world.getJointCount() 
	                       + ", world.getJoints().size -> " + getJointsListSize());
	    
	    world.destroyBody(bodyA);
	    Gdx.app.log("debug", "Destroyed the body A");
	    Gdx.app.log("debug", "world.getJointCount() -> " + world.getJointCount() 
	                       + ", world.getJoints().size -> " + getJointsListSize());
	    
	    world.destroyBody(bodyB);
	    Gdx.app.log("debug", "Destroyed the body B");
	    Gdx.app.log("debug", "world.getJointCount() -> " + world.getJointCount() 
	                       + ", world.getJoints().size -> " + getJointsListSize());
	    
	}
	
	private Body createBody() {
	    BodyDef bodyDef = new BodyDef();
	    return world.createBody(bodyDef);
	}
	
	private void createDistanceJointBetweenBodies(Body bodyA, Body bodyB) {
	    DistanceJointDef jointDefA = new DistanceJointDef();
	    jointDefA.bodyA = bodyA;
	    jointDefA.bodyB = bodyB;
	    world.createJoint(jointDefA);
	}
	
	private int getJointsListSize() {
        Array<Joint> joints = new Array<Joint>();
        world.getJoints(joints);
        return joints.size;
	}
}
```

On desktop, the joints list is empty after destroying one of the bodies, which is the expected behaviour. The application log output:

```
debug: Created the bodies and the joints
debug: world.getJointCount() -> 2, world.getJoints().size -> 2
debug: Destroyed the body A
debug: world.getJointCount() -> 0, world.getJoints().size -> 0
debug: Destroyed the body B
debug: world.getJointCount() -> 0, world.getJoints().size -> 0
```

In the html5 client, however, the list still contains one joint, even though the ```World.getJointCount()``` method returns zero. The application log output:
```
debug: Created the bodies and the joints
debug: world.getJointCount() -> 2, world.getJoints().size -> 2
debug: Destroyed the body A
debug: world.getJointCount() -> 0, world.getJoints().size -> 1
debug: Destroyed the body B
debug: world.getJointCount() -> 0, world.getJoints().size -> 1
```

The problem only occurs when a body with at least two attached joints is destroyed.

I found the similar issue #1913, where a single joint wasn't properly destroyed together with its parent body in GWT. However, that was with only one joint and that issue was closed in May 2014, so this seems to be a different problem

Tested with libGDX 1.7.2 in Chrome 45.0.2454.85 (64-bit) on Linux.

**Comment by MobiDevelop** [comment]
Digging into this a bit and it looks like the createJoint in JBox2D's World is not properly setting up the linked list, so when we iterate over the joint edges in that list to destroy the joints, it is only finding the first.

Not sure why it isn't being set up correctly, so will have to do some digging.

**Comment by Snack-a-Dog** [comment]
Fix confirmed with 1.7.3-SNAPSHOT. Thanks for the super-fast fix :-)

**Comment by loogie** [comment]
Hello!

 

You have a new message, please read <http://www.amydis.com/obvious.php?kbl1m>

 

loogie@basicelementdesign.com
